### ISSUE TITLE
branch: axon-array

### ISSUE THREAD

**Issue by samreid** [phetsims/natural-selection#178]
For https://github.com/phetsims/axon/issues/314, I converted Natural Selection to use AxonArray instead of ObservableArray, including phetioState.  It appears to be working properly in the state wrapper.

**Comment by samreid** [comment]
I committed the branch and it is ready for review.  To explain context: I tried to use Charges and Fields as a test case for AxonArray state, but it wasn't a good use case, so I switched to Natural Selection instead.  I added 4 methods to AxonArray for compatibility with ObservableArray, so most of the usage in Natural Selection was straightforward.  Next step for this issue is to review the changes in Natural Selection and AxonArray.  After review, this could be merged to master.

This does not need to be done before publication.  It will be up to @pixelzoom to decide the best time to work on it.

**Comment by pixelzoom** [comment]
I review these changes. Everything appears to work, and I like this direction. 

Observations related to upcoming publication of this sim:

* This changes the PhET-iO API.   Instead of arrays of type `ObservableArrayIO` they are of type `AxonArrayStateIO`.  That needs to be approved by the NS design team.  

* `AxonArrayStateIO` is a new pattern (delegate) for instrumenting a class that can't use inheritance.  The implementation looks clean in this case.  But where the client might expect to see "IO Type: AxonArrayIO", they see "IO Type: AxonArrayStateIO", which could be a little confusing. Has this pattern been discussed amongst the iO team? Would a PhetioObject mixin be a better longterm solution?  I'm hesitant to expose this in the PhET-iO API until this is solid. 

* The consensus in https://github.com/phetsims/axon/issues/321 is that AxonArray and AxonArrayStateIO are unlikely to be the final names.  So I think it's premature to expose these names in the PhET-iO API, when we are certain to change them and break the API.

* Now that @samreid has added missing methods to AxonArray, this change would be relatively straightforward to do from scratch in the future, probably < 30 minutes. 

Given the higher priorities of PhET-iO performance for this sim, I'm inclined not to include AxonArray in the release of NS 1.0, and to defer until convert other sims.

**Comment by samreid** [comment]
I'll discuss with the PhET-iO subteam on Wednesday about `AxonArrayStateIO`.  I trust your judgment about timing for this simulation.  

> I'm inclined not to include AxonArray in the release of NS, and to defer until we convert other sims.

One possibility would be to wait until we have steady SHAs for the NS RC or a version 1.0, then to work on NS while it is fresh in our minds (instead of having another sim be the first use case).

**Comment by pixelzoom** [comment]
> I'll discuss with the PhET-iO subteam on Wednesday about `AxonArrayStateIO`. 

OK.  But like I said, NS performance is much more important.  If you have limited time, I'd rather see progress on PhET-iO optimizations related to https://github.com/phetsims/natural-selection/issues/140  and https://github.com/phetsims/natural-selection/issues/60.

**Comment by samreid** [comment]
Thanks, I'll make sure those are on the agenda at a higher priority.

**Comment by samreid** [comment]
> Has this pattern been discussed amongst the iO team? Would a `PhetioObject` mixin be a better longterm solution?  I'm hesitant to expose this in the PhET-iO API until this is solid. 

@zepumph @chrisklus and I discussed this today.  @zepumph thought it could just take 1/4 of a day to try porting PhetioObject to be a mixin.  I was skeptical that it could be done so quickly, but even if it could be done quickly, I'm hesitant to use mixins instead of standard inheritance because static analysis doesn't work for mixins (even with ES6 pattern).  For instance, autocomplete, navigation, highlighting does not work for mixins, so I argued that we may prefer the proposed composite pattern anyways.

**Comment by pixelzoom** [comment]
Good points about the disadvantages of mixins.  I've copied that info to https://github.com/phetsims/axon/issues/314#issuecomment-676704021, so it's available in the issue where you're working on that aspect of AxonArray.

Re how to move forward with this branch...  I discussed with @ariel-phet and he agreed that it would be best to defer use of AxonArray in NS, for the reasons listed in https://github.com/phetsims/natural-selection/issues/178#issuecomment-675162248.

While this branch was a nice feasibility test and good for discussion, I don't see it being particularly useful going forward, since it's not going be merged to master. So I suggest that we delete it. Is that OK with you @samreid?

**Comment by samreid** [comment]
It has been a useful test harness as I make changes to AxonArray.  I recommend leaving it until we have another use case, then I agree it can be deleted.

**Comment by pixelzoom** [comment]
8/20/2020 phet-io meeting:

We discussed whether AxonArray should be used in the first release of NS.  @kathy-phet confirmed "no".

**Comment by pixelzoom** [comment]
> It has been a useful test harness as I make changes to AxonArray. I recommend leaving it until we have another use case, then I agree it can be deleted.

Labeling as deferred to review in 3 months or so.

**Comment by pixelzoom** [comment]
AxonArray has been replaced by createArrayProxy, and NS has been converted to createArrayProxy - see #252.   So I believe that this branch can be deleted.  Is that OK with you @samreid?

**Comment by samreid** [comment]
Yes, it is ready to be deleted.  Let me know if you or I should do that.

**Comment by pixelzoom** [comment]
Deleted, closing.

**Comment by pixelzoom** [comment]
Due to https://github.com/phetsims/special-ops/issues/196#issuecomment-823486502, this branch was restored.  It will need to be deleted again.  Reopening and self assigning.

**Comment by pixelzoom** [comment]
Re-deleted using GitHub UI.  Closing.
